### PR TITLE
Formal verification of interrupt detection

### PR DESCRIPTION
**Pull request by francois141**

### CODE REVIEW

**Comment by francois141** [comment]
@CharlyCst You can already review & merge if you want. I still need to make sure that the detected interrupt is the same. Currently I only check if they detects equivalently but I don't check the returned type

File: src/virt/emulator.rs [hunk 1]
```diff
@@ -782,18 +791,34 @@ fn has_user_mode(ctx: &VirtContext) -> bool {
...
+        None
     }
 }
 
+/// Return the ID of the next interrupt to be delivered, if any.
+fn get_next_interrupt(mie: usize, mip: usize, mideleg: usize) -> Option<usize> {
+    // First function
+    let ip = mie & mip & !mideleg;
+
+    // Second function (they should be equivalent)
```
Comment (CharlyCst) [inline src/virt/emulator.rs:818]: What does that mean 'thay should be equivalent'?

File: src/virt/emulator.rs [hunk 2]
```diff
@@ -843,8 +868,7 @@ mod tests {
         assert_eq!(get_next_interrupt(0b000, 0b010, 0b000), None);
         assert_eq!(get_next_interrupt(0b010, 0b010, 0b010), None);
 
-        assert_eq!(get_next_interrupt(0b001, 0b001, 0b000), Some(0));
```
Comment (CharlyCst) [inline src/virt/emulator.rs:846]: This one we can keep, but yes the next one we have to update it with the new order.

File: model_checking/src/symbolic.rs [hunk 3]
```diff
@@ -121,8 +121,9 @@ pub fn new_ctx() -> VirtContext {
     // We fix the architecture type to 64 bits
     ctx.csr.misa = (0b10 << 62) | (ctx.csr.misa & ((1 << 62) - 1));
 
-    // We must have support for usermode in Miralis
+    // We must have support for usermode and in Miralis
```
Comment (CharlyCst) [inline model_checking/src/symbolic.rs:124]: and ? Typo here^^

Is there an issue if there is no S mode? Miralis should support both with and without S mode (but he hardware capabilities must be updated accordingly I think?)

**Review by CharlyCst — COMMENTED** [review]
Here are a few comments.

In addition just to get a better idea, how hard would it be to verify the two functions combined? It seems like you are almost there, but the Sail spec might not have an equivalent function? I am not sure.

File: src/virt/emulator.rs (hunk 1 above)
Comment (francois141) [reply src/virt/emulator.rs:818]: It was for me to debug the formal verification (aka François' brain dump)

File: model_checking/src/symbolic.rs (hunk 3 above)
Comment (francois141) [reply model_checking/src/symbolic.rs:124]: Currently the interrupt virtualisation process won't work if we don't have support for S-mode. Let me handle this in a second PR. The core of the logic is there and can be merged.

**Comment by francois141** [comment]
@CharlyCst I rebased the commit and introduced a new function in Virt-sail which allows us to model check the entire virtualisation of the interrupts at once :)

**Pull request merged**